File: ProblemSolving/450DSA/Python/src/dynamicprogramming/BooleanParenthesization.py
"""
@author Anirudh Sharma

Given a boolean expression S of length N with following symbols.
Symbols
'T' ---> true
'F' ---> false
and following operators filled between symbols
Operators
&   ---> boolean AND
|   ---> boolean OR
^   ---> boolean XOR

Count the number of ways we can parenthesize the expression so that the value of expression
evaluates to true.
"""


def countWays(s):
    # Special case
    if s is None or len(s) == 0:
        return 0
    # Total number of symbols
    n = len(s) // 2 + 1
    # Divide the string into characters of
    # symbols and operators
    symbols = [''] * n
    operators = [''] * (n - 1)
    # Populate these arrays from the input string
    index = 0
    for i in range(0, len(s) - 1, 2):
        symbols[index] = s[i]
        operators[index] = s[i + 1]
        index += 1
    symbols[index] = s[len(s) - 1]
    # Lookup table for making true
    trueLookup = [[0 for y in range(n)] for x in range(n)]
    # Lookup table for making false
    falseLookup = [[0 for y in range(n)] for x in range(n)]
    # Populate the table diagonally
    for gap in range(n):
        i = 0
        for j in range(gap, n):
            # For the zero gap
            if gap == 0:
                if symbols[i] == 'T':
                    trueLookup[i][j] = 1
                else:
                    falseLookup[i][j] = 1
            # For gap greater than one
            else:
                for k in range(i, j):
                    operator = operators[k]
                    # Get different counts for left and
                    # right expressions for both true and false
                    leftTrueCount = trueLookup[i][k]
                    rightTrueCount = trueLookup[k + 1][j]
                    leftFalseCount = falseLookup[i][k]
                    rightFalseCount = falseLookup[k + 1][j]
                    # Solve for different operators
                    # For logical AND operator
                    if operator == '&':
                        trueLookup[i][j] += leftTrueCount * rightTrueCount
                        falseLookup[i][j] += leftTrueCount * rightFalseCount + \
                            leftFalseCount * rightTrueCount + leftFalseCount * rightFalseCount
                    # For logical OR operator
                    elif operator == '|':
                        trueLookup[i][j] += leftTrueCount * rightTrueCount + \
                            leftFalseCount * rightTrueCount + leftTrueCount * rightFalseCount
                        falseLookup[i][j] += leftFalseCount * rightFalseCount
                    # For logical XOR operator
                    else:
                        trueLookup[i][j] += leftTrueCount * \
                            rightFalseCount + leftFalseCount * rightTrueCount
                        falseLookup[i][j] += leftFalseCount * \
                            rightFalseCount + leftTrueCount * rightTrueCount
            i += 1
    return trueLookup[0][n - 1]


if __name__ == "__main__":
    s = "T|T&F^T"
    print(countWays(s))

    s = "T^F|F"
    print(countWays(s))

File: ProblemSolving/450DSA/JavaScript/src/dynamicprogramming/BooleanParenthesization.js
/**
 * @author Anirudh Sharma
 * 
 * Given a boolean expression S of length N with following symbols.
 * Symbols
 * 'T' ---> true
 * 'F' ---> false
 * and following operators filled between symbols
 * Operators
 * &   ---> boolean AND
 * |   ---> boolean OR
 * ^   ---> boolean XOR
 * 
 * Count the number of ways we can parenthesize the expression so that the value of expression
 * evaluates to true.
 */
const countWays = (s) => {
    // Special case
    if (s === undefined || s.length === 0) {
        return 0;
    }
    // Total number of symbols
    const n = parseInt(s.length / 2) + 1;
    // Divide the string into characters of
    // symbols and operators
    const symbols = new Array(n);
    const operators = new Array(n);
    // Populate these arrays from the input string
    let index = 0;
    for (let i = 0; i < s.length - 1; i = i + 2) {
        symbols[index] = s.charAt(i);
        operators[index] = s.charAt(i + 1);
        index++;
    }
    symbols[index] = s.charAt(s.length - 1);
    // Lookup table for making true
    const trueLookup = Array.from(Array(n), () => Array(n).fill(0));
    // Lookup table for making false
    const falseLookup = Array.from(Array(n), () => Array(n).fill(0));
    // Populate the table diagonally
    for (let gap = 0; gap < n; gap++) {
        for (let i = 0, j = gap; j < n; i++, j++) {
            // For the zero gap
            if (gap === 0) {
                if (symbols[i] === 'T') {
                    trueLookup[i][j] = 1;
                } else {
                    falseLookup[i][j] = 1;
                }
            }
            // For gap greater than one
            else {
                for (let k = i; k < j; k++) {
                    let operator = operators[k];
                    // Get different counts for left and
                    // right expressions for both true and false
                    let leftTrueCount = trueLookup[i][k];
                    let rightTrueCount = trueLookup[k + 1][j];
                    let leftFalseCount = falseLookup[i][k];
                    let rightFalseCount = falseLookup[k + 1][j];
                    // Solve for different operators
                    // For logical AND operator
                    if (operator === '&') {
                        trueLookup[i][j] += leftTrueCount * rightTrueCount;
                        falseLookup[i][j] += leftTrueCount * rightFalseCount
                            + leftFalseCount * rightTrueCount
                            + leftFalseCount * rightFalseCount;
                    }
                    // For logical OR operator
                    else if (operator === '|') {
                        trueLookup[i][j] += leftTrueCount * rightTrueCount + leftFalseCount * rightTrueCount + leftTrueCount * rightFalseCount;
                        falseLookup[i][j] += leftFalseCount * rightFalseCount;
                    }
                    // For logical XOR operator
                    else {
                        trueLookup[i][j] += leftTrueCount * rightFalseCount + leftFalseCount * rightTrueCount;
                        falseLookup[i][j] += leftFalseCount * rightFalseCount + leftTrueCount * rightTrueCount;
                    }
                }
            }
        }
    }
    return trueLookup[0][n - 1];
};

const main = () => {
    let S = "T|T&F^T";
    console.log(countWays(S));

    S = "T^F|F";
    console.log(countWays(S));
};

main();